airflow/models.py

```diff
@@ -4398,13 +4397,20 @@ def update_state(self, session=None):
         # small speed up
         if unfinished_tasks and none_depends_on_past:
             # todo: this can actually get pretty slow: one task costs between 0.01-015s
-            no_dependencies_met = all(
-                # Use a special dependency context that ignores task's up for retry
-                # dependency, since a task that is up for retry is not necessarily
-                # deadlocked.
-                not t.are_dependencies_met(dep_context=self.DEADLOCK_CHECK_DEP_CONTEXT,
-                                           session=session)
-                for t in unfinished_tasks)
+            no_dependencies_met = True
+            for ut in unfinished_tasks:
+                # We need to flag upstream and check for changes because upstream
+                # failures can result in deadlock false positives
+                old_state = ut.state
+                deps_met = ut.are_dependencies_met(
+                    dep_context=DepContext(
+                        flag_upstream_failed=True,
+                        ignore_in_retry_period=True),
+                    session=session)
+                new_state = ut.current_state(session=session)
```

Comment:
We need this check because `flag_upstream_failed` will change the state of tasks but still return a fail condition.

There is an issue that the dependency model only has True/False for dependencies being met but upstream_failed and upstream skipping are cases where the dependency isn't met per se but that we don't want to count it against the dag moving forward.

Basically, the task is blocked but the DAG is not, which is a level of granularity the current boolean return doesn't account for.

Longer term, it might make sense to have a richer possible return state for dependency checks but, in the short term resolving the deadlock issue was my priority.